### PR TITLE
FIX-2009 Update logCurveInfo properties

### LINKED ISSUE #2009: User story: Frontend - Update logCurveInfo attributes in property window
As a Data Manager, I want to be able to right click on a curve/mnemonic, and update the attributes in the property window.

NB: This backlog Item only covers single instance update, not batch update of multiple object instances):

Acceptance criteria
 Ability to right click on a curve/mnemonic Object instance, and update the curve/mnemonic attributes listed below:

image
Example above how it looks today, we also would like:

**unit - editable
curveDescription - editable
typeLogData - uneditable
mnemAlias - uneditable**

### PR DESCRIPTION
**Pull request by jaroslavbliznak**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes #2009 

## Description

[//]: # (Please include a written summary of the changes and which issue is fixed. List any dependencies that are required for this change._)

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [x] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [X] Frontend
* [X] API
* [ ] WITSML
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [X] I have self-reviewed my code
* [X] No new warnings are generated

*Test coverage*
* [X] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: Tests/WitsmlExplorer.IntegrationTests/Api/Workers/RenameMnemonicWorkerTests.cs [hunk 1]
```diff
@@ -8,14 +8,15 @@
 using WitsmlExplorer.Api.Models;
```
Comment (janmarius) [inline Tests/WitsmlExplorer.IntegrationTests/Api/Workers/RenameMnemonicWorkerTests.cs:8]: Do we also need to change the name of this file?

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            {
+                const string errorMessage = "Failed to modify logCurveInfo object.";
+                Logger.LogError("{ErrorMessage}. {jobDescription}", errorMessage, job.Description());
+                return (new WorkerResult(GetTargetWitsmlClientOrThrow().GetServerHostname(), false, errorMessage, modifyLogCurveInfoResult.Reason), null);
+            }
+
+            if (originalLogCurveInfoMnemonic != job.LogCurveInfo.Mnemonic)
+            {
+                var originalMnemonics = GetMnemonics(logHeader, originalLogCurveInfoMnemonic);
+                var updatedMnemonics = string.Join(",", GetMnemonics(logHeader, job.LogCurveInfo.Mnemonic));
```
Comment (janmarius) [inline Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: Is it appropriate to use `DataSeparator` found in `CommonConstants.cs` in this context?

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                var updatedUnits = string.Join(",", GetUnits(logHeader, job.LogCurveInfo.Unit));
```
Comment (janmarius) [inline Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: `DataSeparator`?

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+            if (string.IsNullOrEmpty(job.LogCurveInfo.Unit))
+            {
+                throw new InvalidOperationException("An empty unit is given when trying to change a unit. Make sure a valid unit is given.");
+            }
+        }
+
+        private static List<string> GetUnits(WitsmlLog log, string newUnit)
+        {
+            var indexCurveUnit = log.LogCurveInfo.FirstOrDefault(x => x.Mnemonic == log.IndexCurve.Value)?.Unit ?? CommonConstants.DefaultUnit;
```
Comment (janmarius) [inline Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: Has the team discussed and approved setting non-existing units to 'unitless' when no unit is provided for the mnemonic?

File: Src/Witsml/CommonConstants.cs [hunk 5]
```diff
@@ -8,4 +8,5 @@ public static class CommonConstants
     public const int DefaultNumberOfRoundedPlaces = 3;
     public const string DataSeparator = ",";
     public const string NewLine = "\n";
+    public const string DefaultUnit = "unitless";
```
Comment (janmarius) [inline Src/Witsml/CommonConstants.cs]: Can you please confirm if this is the team's approved default behavior?

File: Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            _witsmlClient.Setup(client =>
+                    client.UpdateInStoreAsync(It.IsAny<WitsmlLogs>())).Callback<WitsmlLogs>(logs => updatedLogs.Add(logs))
+                .ReturnsAsync(new QueryResult(true));
+
+            await _worker.Execute(job);
+
+            Assert.Single(updatedLogs);
+            Assert.Equal(expectedMnemonic, updatedLogs.First().Logs.First().LogCurveInfo.First().Mnemonic);
+            Assert.Equal(expectedUnit, updatedLogs.First().Logs.First().LogCurveInfo.First().Unit);
+            Assert.Equal(expectedCurveDescription, updatedLogs.First().Logs.First().LogCurveInfo.First().CurveDescription);
```
Comment (janmarius) [inline Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs]: Would using `FirstOrDefault()` be a better approach in this particular situation, considering that it does not throw any errors?

File: Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                .ReturnsAsync(new QueryResult(true));
+
+            _witsmlClient.Setup(client =>
+                    client.DeleteFromStoreAsync(It.IsAny<WitsmlLogs>()))
+                .ReturnsAsync(new QueryResult(true));
+
+            await _worker.Execute(job);
+
+            Assert.Single(updatedLogs);
+            Assert.Equal(expectedMnemonic, updatedLogs.First().Logs.First().LogCurveInfo.First().Mnemonic);
```
Comment (janmarius) [inline Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs]: FirstOrDefault?

File: Src/WitsmlExplorer.Frontend/components/Modals/LogCurveInfoPropertiesModal.tsx [hunk 8]
```diff
@@ -51,11 +49,27 @@ const LogCurveInfoPropertiesModal = (props: LogCurveInfoPropertiesModalProps): R
...
                 fullWidth
                 inputProps={{ minLength: 1, maxLength: 64 }}
                 onChange={(e) => setEditableLogCurveInfo({ ...editableLogCurveInfo, mnemonic: e.target.value })}
               />
+              <TextField
+                id="unit"
+                label="unit"
+                defaultValue={editableLogCurveInfo.unit}
+                error={editableLogCurveInfo.unit == null || editableLogCurveInfo.unit.length === 0}
+                helperText={editableLogCurveInfo.unit == null || editableLogCurveInfo.unit.length === 0 ? "A unit cannot be empty." : ""}
```
Comment (janmarius) [inline Src/WitsmlExplorer.Frontend/components/Modals/LogCurveInfoPropertiesModal.tsx]: Doesn't a unit have a maximum length?

**Review by janmarius — CHANGES_REQUESTED** [review]
I added suggestions and comments on minor issues. During testing, I also noticed changing the index curve was possible. Is this expected behavior?

File: Tests/WitsmlExplorer.IntegrationTests/Api/Workers/RenameMnemonicWorkerTests.cs (hunk 1 above)
Comment (jaroslavbliznak) [reply Tests/WitsmlExplorer.IntegrationTests/Api/Workers/RenameMnemonicWorkerTests.cs:8]: Renamed.

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs (hunk 2 above)
Comment (jaroslavbliznak) [reply Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: Fixed.

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs (hunk 3 above)
Comment (jaroslavbliznak) [reply Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: Fixed.

File: Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs (hunk 4 above)
Comment (jaroslavbliznak) [reply Src/WitsmlExplorer.Api/Workers/Modify/ModifyLogCurveInfoWorker.cs]: I discussed it with Elias. I removed 'unitless' and in this case, it is relevant not to specify the default unit. (It means an empty value will be used.)

File: Src/Witsml/CommonConstants.cs (hunk 5 above)
Comment (jaroslavbliznak) [reply Src/Witsml/CommonConstants.cs]: DefaultUnit was removed.

File: Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs (hunk 6 above)
Comment (jaroslavbliznak) [reply Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs]: Fixed.

File: Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs (hunk 7 above)
Comment (jaroslavbliznak) [reply Tests/WitsmlExplorer.Api.Tests/Workers/ModifyLogCurveInfoWorkerTests.cs]: Fixed.

File: Src/WitsmlExplorer.Frontend/components/Modals/LogCurveInfoPropertiesModal.tsx (hunk 8 above)
Comment (jaroslavbliznak) [reply Src/WitsmlExplorer.Frontend/components/Modals/LogCurveInfoPropertiesModal.tsx]: Fixed.

**Comment by jaroslavbliznak** [comment]
I have fixed the modification if it is the index curve. In this case, it is not possible to rename the mnemonic.

**Review by janmarius — CHANGES_REQUESTED** [review]
I am unable to change the name of the mnemonic (error shown in image below). Additionally, it is still possible to modify the unit of the index curve. Should this be allowed?

image

**Comment by jaroslavbliznak** [comment]
This is an error with overlapping indexes during editing, and it only occurs on the test server for Petrolink (https://eqwitsml.demo.petrolink.net/Petrovault/WITSML/WMLS.asmx). Asbjørn has already reported a message about it to them. It should work fine on other servers. The unit should be editable for the index curve as well.

**Review by janmarius — APPROVED** [review]
*Reviewer checklist*
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code

**Pull request merged**